### PR TITLE
ElkUtil: Filter unnecessary bend points in Junction Point calculation

### PR DESCRIPTION
**Pull request by eNBeWe**
When junction points are generated in fixed layout, unnecessary bend
points might lead to misplaced junction points. This obviously not
wanted.

misplacedjunction

```
edgeRouting: ORTHOGONAL
algorithm: fixed

node N1 {
	layout [ 
		position: 0, 0 
		size: 20, 20
	]

	port P1 {
		layout [
			position: 20, 7.5
			size: 5, 5
		]
	}
}

node N2 {
	layout [
		position: 150, 0
		size: 20, 20
	]
}

node N3 {
	layout [
		position: 100, 50
		size: 20, 20
	]

}

edge N1.P1 -> N2 {
	layout [
		start: 20, 10
		end: 150, 10
		bends: 40, 10
	]
}

edge N1.P1 -> N3 {
	layout [
		start: 20, 10
		end: 100, 60
		bends: 80, 10 | 80, 60
	]
}
```

### CODE REVIEW

File: plugins/org.eclipse.elk.core/src/org/eclipse/elk/core/util/ElkUtil.java [hunk 1]
```diff
@@ -626,28 +626,43 @@ public static KVector toRelative(final KVector point, final ElkNode parent) {
...
+
+        // Filter unnecessary bend points from the list
+        int i = 1;
+        while (i < points.size() - 1) {
+            // Unnecessary bend points are given if three points in a row have the same x or y coordinate
+            KVector p1 = points.get(i - 1);
+            KVector p2 = points.get(i);
+            KVector p3 = points.get(i + 1);
+            
+            if ((p1.x == p2.x && p2.x == p3.x) || (p1.y == p2.y && p2.y == p3.y)) {
```
Comment (uruuru) [inline plugins/org.eclipse.elk.core/src/org/eclipse/elk/core/util/ElkUtil.java]: Equality on doubles may be problematic.

File: plugins/org.eclipse.elk.core/src/org/eclipse/elk/core/util/ElkUtil.java [hunk 2]
```diff
@@ -626,28 +626,43 @@ public static KVector toRelative(final KVector point, final ElkNode parent) {
...
+                // Found a straight segment, drop p2 and re-check with the same i
+                points.remove(i);
```
Comment (uruuru) [inline plugins/org.eclipse.elk.core/src/org/eclipse/elk/core/util/ElkUtil.java]: Remove on an ArrayList is inefficient.

**Comment by eNBeWe** [comment]
The unnecessary bend points are a problem because the junction point calculation moves along all edges originating from one port in lockstep. An extra bend point in one edge messes this up.

This lockstep also makes modifying the computation step way more complicated than this solution, requiring to keep individual pointers for all edges currently following the "main" edge path.

File: plugins/org.eclipse.elk.core/src/org/eclipse/elk/core/util/ElkUtil.java (hunk 2 above)
Comment (eNBeWe) [reply plugins/org.eclipse.elk.core/src/org/eclipse/elk/core/util/ElkUtil.java]: I am well aware of that, but the removing part should be the rare case, compared to the more or less random access during iteration.

File: plugins/org.eclipse.elk.core/src/org/eclipse/elk/core/util/ElkUtil.java (hunk 1 above)
Comment (eNBeWe) [reply plugins/org.eclipse.elk.core/src/org/eclipse/elk/core/util/ElkUtil.java]: Sure, but
a) we are talking about fixed layouter, meaning the coordinates are passed by the model and not some calculation that is likely to create small deltas
b) if there is an offset the edge isn't entirely straight, is it?

**Pull request merged**